junit5/src/main/java/cz/xtf/junit5/listeners/ProjectCreator.java

```diff
@@ -1,92 +1,76 @@
 package cz.xtf.junit5.listeners;
 
-import java.util.concurrent.TimeUnit;
-import java.util.function.BooleanSupplier;
+import static cz.xtf.core.config.OpenShiftConfig.OPENSHIFT_NAMESPACE;
 
+import java.util.Arrays;
+
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.engine.descriptor.MethodBasedTestDescriptor;
+import org.junit.platform.engine.FilterResult;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.launcher.PostDiscoveryFilter;
 import org.junit.platform.launcher.TestExecutionListener;
 import org.junit.platform.launcher.TestPlan;
 
 import cz.xtf.core.config.OpenShiftConfig;
-import cz.xtf.core.openshift.OpenShift;
+import cz.xtf.core.config.XTFConfig;
+import cz.xtf.core.context.TestCaseContext;
+import cz.xtf.core.namespace.NamespaceManager;
 import cz.xtf.core.openshift.OpenShifts;
-import cz.xtf.core.waiting.SimpleWaiter;
 import cz.xtf.junit5.config.JUnitConfig;
-import io.fabric8.kubernetes.api.builder.Visitor;
-import io.fabric8.kubernetes.api.model.NamespaceBuilder;
-import io.fabric8.kubernetes.client.KubernetesClientException;
 import lombok.extern.slf4j.Slf4j;
 
 @Slf4j
-public class ProjectCreator implements TestExecutionListener {
-    private static final OpenShift openShift = OpenShifts.master();
+public class ProjectCreator
+        implements TestExecutionListener, BeforeAllCallback, AfterAllCallback, PostDiscoveryFilter {
 
     @Override
     public void testPlanExecutionStarted(TestPlan testPlan) {
-        createProject();
+        NamespaceManager.createIfDoesNotExistsProject(XTFConfig.get(OPENSHIFT_NAMESPACE));
     }
 
     @Override
-    public void testPlanExecutionFinished(TestPlan testPlan) {
-        if (JUnitConfig.cleanOpenShift()) {
-            deleteProject();
-        }
+    public void beforeAll(ExtensionContext context) {
+        // todo consider to move setting testcase context in separate extension
```

Comment:
This is good point. Especially when you say there si at least one team which does not use ProjectCreator.

I think such new Extension could be registered by default.

Reply:
I was looking at and it does not seems to be possible to register extension by default in consuming test suite.

Reply:
Well, looks like it can be automatically added if consuming test suite has:
`junit.jupiter.extensions.autodetection.enabled=true` (be default it's `false`)

In this case consuming test suite does not need to add it in `META-INF/services/org.junit.jupiter.api.extension.Extension` as it can be registered as extension directly in `cz.xtf.junit5` maven depdency.

However even though it seems to be cleaner design, setting `TestCaseContext` in separate extension must happen before `ProjectCreator` is called. I've tried the above appoach and `ProjectCreator` is called first, before `TestCaseContext` is set which leads to `RuntimeException`. 

Only solution seems to have both. Setting it in extension and here.

Reply:
Do you have any opinion on this?

As `TestCaseContext` is public I tend to add this extension into `core` module as it's used there and publicly accesible. So use both - setting in `ProjectCreator` and in extension. wdyt?

Reply:
Introduce own Extension sounds to me correct.

Need to keep it here in ProjectCreator seems to me as workaround. Isn't there way to ensure filling `TestCaseContext` is called as very first code when running TestCase?

Reply:
Or simply document in Readme that to properly configure this feature `TestCaseContextExtension` must be registered first?

Reply:
> Need to keep it here in ProjectCreator seems to me as workaround. Isn't there way to ensure filling TestCaseContext is called as very first code when running TestCase?

Only way seems to be over extension. I didn't find any other way. 

> Or simply document in Readme that to properly configure this feature TestCaseContextExtension must be registered first?

I'll document it but I think having it here will avoid user issues wondering why it's not working. 

I've added it to both places.

Reply:
I do not see it documented. Am I missing something?

Reply:
Unfortunately `ProjectCreator` is registered and invoked first so it does not solve the problem.